Update payment request when billing data changes

reykjalin opened the pull request:
Fixes #1830 


# Changes proposed in this Pull Request:

- Update the `PaymentRequest` object when billing data changes in the cart block instead of recreating it.

The `PaymentRequestButtonElement` doesn't support re-assigning the `paymentRequest` prop, and as such we have to update the payment request instead.

We achieve this by sending an API request for the cart information to make sure all normalizations are correctly applied, and use that data to update the `PaymentRequest` object.

# Testing instructions

## Make sure Cart Block updates are applied to the Payment Request

> **Note:** This test requires that you have the [WooCommerce Blocks](https://wordpress.org/plugins/woo-gutenberg-products-block/) extension installed on your store.

1. Add a product to your cart.
2. Go to the page containing the Cart Block.
3. Click the Payment Request Button (Apple Pay/Google Pay/Chrome Payment Request).
4. Note the list of items and the total.
5. Click **Cancel** to cancel the payment.
6. Change the number of items in the cart.
7. Click the Payment Request Button again.
8. Make sure the list of items and the total have been updated.

## Make sure Payment Requests can be cancelled multiple times in the Cart Block

> **Note:** This test requires that you have the [WooCommerce Blocks](https://wordpress.org/plugins/woo-gutenberg-products-block/) extension installed on your store.

1. Add a product to your cart.
2. Go to the page containing the Cart Block.
3. Click the Payment Request Button (Apple Pay/Google Pay/Chrome Payment Request).
4. Click **Cancel** to cancel the payment.
5. Click the Payment Request Button again.
6. Click **Cancel** to cancel the payment.
7. Change the number of items in the cart.
8. Click the Payment Request Button again.
9. Click **Cancel** to cancel the payment.
10. Click the Payment Request Button again.
11. Click **Cancel** to cancel the payment.
12. Make sure you can interact with the Payment Request Button (i.e. button is clickable, and not greyed out).
13. Change the number of items in the cart.
14. Again, make sure the Payment Request Button is still active (i.e. button is clickable, and not greyed out).

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

Commit:
Update payment request when billing data changes

Applies only to the Cart Block integration.

The PaymentRequestButtonElement doesn't support re-assigning the
paymentRequest prop, and as such we have to update the payment request
instead.

We achieve this by sending another API request for the cart to make sure
all normalizations are correctly applied, and use that data to update
the payment request itself.

diegocurbelo reviewed (approved):
The code changes look good and both cases test well.

Just one small detail... if you click the `Buy now` as soon as is re-enabled you get the following js error in the console, and the amount is not updated, If you wait 1 or 2 seconds it works fine; maybe is my test env, and is not a real-life issue just wanted to rise the issue just in case
<img width="795" alt="Screen Shot 2021-10-19 at 18 37 49" src="https://user-images.githubusercontent.com/407542/137994441-855c7a72-c5c3-4fa7-9a09-51ac8ff1f56e.png">

Commit:
Merge branch 'develop' into fix/1830-cart-block-updates-propagate-to-payment-request

Commit:
Remove unnecessary tslint directives

reykjalin commented:
Nice catch @diegocurbelo ! I'm looking into how we can disable the PRB until the cart request has completed, but wasn't able to figure it out on my own. I'll ask Team Rubik in p1634710351348900-slack-C8X6Q7XQU and see if they can help us 👍

Commit:
Merge branch 'develop' into fix/1830-cart-block-updates-propagate-to-payment-request

Commit:
Disable PRB while the payment request is updating

We had to do this in a pretty hacky way since there is no way to disable
Stripe's PaymentRequestButton Element.

Hopefully we'll get some way to control this through the Blocks API.

reykjalin commented:
@diegocurbelo I've fixed the issue you found, but the fix is a bit hacky. I couldn't figure out a better way to accomplish this. I'm very open to ideas — there's also more discussion in the Team Rubik thread in Slack 🙂

diegocurbelo commented on client/blocks/payment-request/payment-request-express.js:
```diff
@@ -120,15 +121,31 @@ const PaymentRequestExpressComponent = ( {
+					paymentRequest,
+				} }
+			/>
+		</div>
```
How about doing something like this? (very very hacky I know 😞)...

The benefit of this approach is that we keep the button "disabled" until is clickable... I know is just 1 second and maybe not worth it, so feel free to disregard this suggestion

```suggestion
		<div
			{ ...( isUpdatingPaymentRequest && {
				className: 'wc-block-components-loading-mask',
			} ) }
		>
			<div
				{ ...( isUpdatingPaymentRequest && {
					className: 'wc-block-components-loading-mask__children',
				} ) }
			>
				<PaymentRequestButtonElement
					onClick={ onPaymentRequestButtonClick }
					options={ {
						style: paymentRequestButtonStyle,
						paymentRequest,
					} }
				/>
			</div>
		</div>
```

reykjalin replied:
Works like a charm, although I added the child element class directly onto the Stripe Element itself 🙂 

Fixed in afa9bf6b259ebaa4324d023965a8d7f092af15c1!

reykjalin replied:
Nice suggestion by the way @diegocurbelo ! I couldn't figure out the necessary class combinations for this 👏

Commit:
Add divs around the PRBs to trigger loading state

We only trigger the specific classes necessary to force the loading
state while the updated cart is being retrieved to update the payment
request.

diegocurbelo committed:
Merge branch 'develop' into fix/1830-cart-block-updates-propagate-to-payment-request

diegocurbelo reviewed (approved):
Works perfectly @reykjalin... good job! 🚢 🚢

Commit:
Merge branch 'develop' into fix/1830-cart-block-updates-propagate-to-payment-request

Commit:
add changelog entry

reykjalin commented:
Added a changelog entry, and added the testing instructions to the [5.8.0 release testing instructions](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-testing-instructions-for-the-WooCommerce-Stripe-payment-gateway-5.8.0). I'll merge when the tests have succeeded 🎉

asumaran commented:
@reykjalin the problem this PR fixes is very similar to (one of) the issues I've been trying to solve on PRBs (pcrvl0-e2-p2).

This approach will work just fine for simple products but I'm sure it won't work with variable products with different shipping types (normal/virtual) since we can't update the PRB to change the `requestShipping` option once the PRB is created. The only way to do that -I found so far- is to destroy the button and create it again with `requestShipping` the updated value. 

I just tested (develop 5d841445) with a product with two variations. One requires shipping and the other one doesn't. When the item that requires shipping is removed the payment sheet still asks for a Shipping Address even though the only added product is a virtual product which doesn't require shipping therefore a shipping address should not be required.

<img width="903" alt="Screen Shot 2021-10-26 at 11 15 17 PM" src="https://user-images.githubusercontent.com/1025173/138998926-ee721850-f0a0-4f66-959e-ab6efda425b9.png">
